clients/drcachesim/scheduler/scheduler.h

```diff
@@ -0,0 +1,562 @@
+         * occupied by the input stream set.
+         * The only stream type supported here is
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::STREAM_BY_RECORDED_CPU.
+         */
+        SCHEDULE_INTERLEAVE_AS_RECORDED,
+        // TODO i#5843: Can we get doxygen to link references without the
+        // fully-qualified name?  These are so long I had to omit the
+        // 'quantum_duration' to stay in the vera++-checked line limit.
+        /**
+         * Uses a synthetic schedule with pre-emption of an input to be replaced
+         * by another input on a given output stream when it uses up its
+         * quantum which is measured using instruction counts.
+         * The quantum size is specified by the `quantum_duration` field of
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::scheduler_options_t.
+         */
+        SCHEDULE_BY_INSTRUCTION_QUANTUM,
+        /**
+         * Uses a synthetic schedule with pre-emption of an input to be replaced
+         * by another input on a given output stream when it uses up its
+         * quantum which is measured using the user's notion of time.
+         * This notion of time must be supplied by the user by calling
+         * dynamorio::drmemtrace::scheduler_tmpl_t::stream_t::report_time()
+         * periodically.
+         * The quantum size is specified by the `quantum_duration` field of
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::scheduler_options_t.
+         */
+        SCHEDULE_BY_TIME_QUANTUM,
+    };
+
+    /** Flags controlling aspects of the scheduler beyond the core scheduling. */
+    enum scheduler_flags_t {
+        /** Default behavior. */
+        SCHEDULER_DEFAULTS = 0,
+        // TODO i#5843: Decide and then describe whether the physaddr is in every
+        // record or obtained via an API call.
+        /**
+         * Whether physical addresses should be provided in addition to virtual
+         * addresses.
+         */
+        SCHEDULER_PROVIDE_PHYSICAL_ADDRESSES = 0x1,
+        /**
+         * Specifies that speculation should supply just NOP instructions.
+         */
+        SCHEDULER_SPECULATE_NOPS = 0x2,
+        // TODO i#5843: Add more speculation flags for other strategies.
+    };
+
+    /**
+     * Collects the parameters specifying how the scheduler should behave, outside
+     * of the workload inputs and the output count.
+     */
+    struct scheduler_options_t {
+        /** Constructs a default set of options. */
+        scheduler_options_t()
+        {
+        }
+        /** Constructs a set of options with the given type and strategy. */
+        scheduler_options_t(stream_type_t stream_type, schedule_strategy_t strategy)
+            : how_split(stream_type)
+            , strategy(strategy)
+        {
+        }
+
+        /** Size of the struct for binary-compatible additions. */
+        size_t struct_size = sizeof(scheduler_options_t);
+        /** The base mode of the scheduler with respect to input streams. */
```

Comment:
Usage with the analyzer will explicitly set the mode to analyzer models; usage on a direct scheduler is likely to be a simulator and not an analysis tool and this mode seems the default you'd want there (though it is not yet implemented it will be soon).